extract/spotify_api.py

```diff
@@ -0,0 +1,239 @@
+    result = json.loads(result.content)
+    return result["items"]
+
+
+def create_track_dicts(items: list[dict]) -> list[dict]:
+    """Takes in a list of playlist items and returns a list of dictionaries of each track"""
+    tracks = []
+    rank = 0
+    for item in items:
+        rank += 1
+        track = {}
+        track["id"] = item["track"]["id"]
+        track["name"] = item["track"]["name"]
+        track["spotify_rank"] = rank
+        track["tiktok_rank"] = None
+        track["in_spotify"] = True
+        track["in_tiktok"] = False
+        track["popularity"] = get_track_popularity(track["id"], headers)
+        audio_features = get_track_audio_features(track["id"], headers)
+        track["danceability"] = audio_features[0]
+        track["energy"] = audio_features[1]
+        track["valence"] = audio_features[2]
+        track["tempo"] = audio_features[3]
+        track["speechiness"] = audio_features[4]
+
+        track["artists"] = []
+        for artist in item["track"]["artists"]:
+            artist_dict = {}
+            artist_dict["id"] = artist["id"]
+            artist_dict["name"] = artist["name"]
+            artist_followers = get_artist_followers(artist["id"], headers)
+            artist_dict["popularity"] = artist_followers[0]
+            artist_dict["follower_count"] = artist_followers[1]
+            artist_dict["genres"] = artist_followers[2]
+            track["artists"].append(artist_dict)
+        tracks.append(track)
+    return tracks
+
+
+def get_artist_followers(artist_id: str, headers: dict) ->tuple:
+    """Takes an artist id and gets the popularity rating and follower count for that artist"""
+    result = get(f"{SPOTIFY_BASE_URL}artists/{artist_id}", headers=headers, timeout=10)
+    result = json.loads(result.content)
+    popularity = result["popularity"]
+    follower_count = result["followers"]["total"]
+    genres = result["genres"]
+    return (popularity, follower_count, genres)
+
+
+def get_track_popularity(track_id: str, headers: dict) -> tuple:
+    """Takes a track id and gets the popularity rating of that track"""
+    result = get(f"{SPOTIFY_BASE_URL}tracks/{track_id}", headers=headers, timeout=10)
+    result = json.loads(result.content)
+    if "popularity" in result:
+        popularity = result["popularity"]
+    else:
+        popularity = None
+    return popularity
+
+
+def get_track_audio_features(track_id: str, headers: dict) -> tuple:
+    """Takes a track id and gets danceability, energy, valence, tempo, and speechiness scores"""
+    result = get(f"{SPOTIFY_BASE_URL}audio-features/{track_id}", headers=headers, timeout=10)
+    result = json.loads(result.content)
+    danceability = result["danceability"]
+    energy = result["energy"]
+    valence = result["valence"]
+    tempo = result["tempo"]
+    speechiness = result["speechiness"]
+    return (danceability, energy, valence, tempo, speechiness)
```

Comment:
Potentially could return these attributes in a dict rather than a tuple to make it more clear when you reference them above rather than indices.

Reply:
I figured that since they were immediately being put into a dict once returned, it would be a bit clunky to put them into a dict, just to immediately take them back out again and put them in another dict